Add support for OpenBSD

buzzdeee opened the pull request:
commit 94a1facf317066cfe266412ebc7cc2a803375526
Author: Sebastian Reitenbach <sebastia@l00-bugdead-prods.de>
Date:   Tue Jun 9 20:01:17 2015 +0200

        The ruby interpreter is /usr/local/bin/ruby21, named after
           the ruby version it's used to run. I made the interpreter configurable
           instead of hardcoding another one in.
           Therefore added a ruby_interpreter parameter, that defaults to
           what was there hardcoded.
           SPEC tests added for the interpreter.

commit 91d0f2e9aa0e09f382d96aa18c69631fefdb2aef
Author: Sebastian Reitenbach <sebastia@l00-bugdead-prods.de>
Date:   Tue Jun 9 20:00:31 2015 +0200

        * there is no distinction between client and server package,
           it's just a single package. For that reason, a check is added
           around client and server package definition, checking if it
           might already be defined. Therefore added server_package and
           client_package parameters, defaulting to what was there hard-coded.
           SPECS still seems to work.

commit 1ec792560312054a1347544fce77bca8d8cab557
Author: Sebastian Reitenbach <sebastia@l00-bugdead-prods.de>
Date:   Tue Jun 9 19:56:16 2015 +0200

         * I use Hiera, to define all the gory parameters, with that there
           was trouble, getting the boolean value of middleware_ssl from there
           into the module, so had to add a check around it in the
           activemq/hosts_iteration.pp and rabbitmq/hosts_iteration.pp
           This is also the reason why I had to bump minimal required stdlib
           version from 3.2.0 to 4.2.0

first time I had to unsquash commits, and it worked without messing up with Git ;)
While there, the original pull requests, due to resetting my fork disappeared :(
Anyways, now hope that meets your expectations.

Otherwise let me know how to improve, and I'll be happy to do so.

cheers,
Sebastian

nibalizer commented on manifests/client/install.pp:
```diff
@@ -5,8 +5,11 @@
-    package { $mcollective::client_package:
-      ensure => $mcollective::version,
+    # prevent conflict where client package name == server package name
+    if !defined(Package[$mcollective::client_package]) {
```
I think it would be cleaner to 
```puppet
# Some systems (OpenBSD) have the same package for both
if $mco::client_package != $mco::server_package { 
  #stuff 
}
```

buzzdeee replied:
indeed makes much more sense

nibalizer commented on manifests/common/config/connector/rabbitmq/hosts_iteration.pp:
```diff
@@ -5,7 +5,12 @@
   }
 
-  $port = $mcollective::middleware_ssl ? {
+  $middleware_ssl = is_bool($mcollective::middleware_ssl) ? {
```
Maybe any2bool?

buzzdeee replied:
going to figure out tomorrow.

nibalizer commented on manifests/init.pp:
```diff
@@ -49,6 +49,7 @@
   $service_name       = 'mcollective',
   $server_package     = 'mcollective',
   $ruby_stomp_package = 'ruby-stomp',
+  $ruby_interpreter   = '/usr/bin/env ruby',
```
Is this pe compatible? There are some other PRs floating around that munge intepreter to be pe safe

buzzdeee replied:
in manifests/server/config/factsource/yaml.pp there is the $ruby_shebang_path set based on $::is_pe, so setting it to something for pe, hardcoded, or now after the patch, to $ruby_interpreter.
That should be pe compatible. The spec tests, I added, also test for the is_pe case.

nibalizer commented on manifests/server/install.pp:
```diff
@@ -5,8 +5,11 @@
-    package { $mcollective::server_package:
-      ensure => $mcollective::version,
+    # prevent conflict where client package name == server package name
+    if !defined(Package[$mcollective::server_package]) {
```
This logic can be removed if the above suggestion is implemented.

igalic commented:
@buzzdeee what's the status here?

buzzdeee commented:
I use it since I opened the PR and it works for me, only that the ruby_interpreter meanwhile changed to be ruby22 ;) But since it's a parameter now, no big deal anymore.

Let me know about this oski dependency for the any2bool, and I might change/add it if you want.

However, I rebased, and things still seem to work for me.

buzzdeee commented:
I rebased to current HEAD, while there, I figured, the cron job for the mcollective-refresh-metadata broke for me on OpenBSD, due to the use of 'bash' as interpreter. I changed that to '/bin/sh', see the added commit. 
I also changed a bit how the ruby_interpreter variable is handled. Now default values are defined based on PE or not in defaults.pp, and can be overridden via the parameter to mcollective. 
In the yaml.pp file, I removed that if/else statement, and only refer to $mcollective::ruby_interpreter. 
That looks a bit cleeaner in my eyes, and additionally allows overriding the ruby interpreter also in PE case, which was not possible before.. Further the spec tests were adopted, and made a bit cleaner too.

btw: I also found a any2bool() function in example42/puppi module, but still the question, is a new
dependency wanted?

Anything else I could do here?

igalic commented:
@buzzdeee thank you for this update, and… sorry for breaking it again._.
i just merged a pr by @alexjfisher which fixed some of the issues you mention

buzzdeee commented:
nevermind, I'm going to rebase, and see how it works out again, that's not dramatic to me ;)

buzzdeee commented:
alright, now rebased again, kept the str2bool around the $::is_pe that he added, so how about that now? At least also his new spec tests are also still green ;)

cheers,
Sebastian

igalic commented on manifests/common/config/connector/activemq/hosts_iteration.pp:
```diff
@@ -6,7 +6,12 @@
 
-  $port = $mcollective::middleware_ssl ? {
+  $middleware_ssl = is_bool($mcollective::middleware_ssl) ? {
+    true  => $mcollective::middleware_ssl,
```
str2bool isn't necessary for true?

alexjfisher commented on .fixtures.yml:
```diff
@@ -9,6 +9,6 @@ fixtures:
     stdlib:
       repo: 'https://github.com/puppetlabs/puppetlabs-stdlib.git'
-      ref: '4.0.0'
+      ref: '4.2.0'
```
for is_bool ?

alexjfisher commented on manifests/common/config/connector/activemq/hosts_iteration.pp:
```diff
@@ -6,7 +6,12 @@
-  $port = $mcollective::middleware_ssl ? {
+  $middleware_ssl = is_bool($mcollective::middleware_ssl) ? {
+    true  => $mcollective::middleware_ssl,
+    false => str2bool($mcollective::middleware_ssl),
```
No need to use is_bool() here.  Since stdlib 4.0.0, str2bool also accepts something that's already a bool and just returns it.

alexjfisher commented on manifests/defaults.pp:
```diff
@@ -30,4 +30,10 @@
+  } else {
+    $ruby_interpreter = '/usr/bin/env ruby'
+  }
+
```
@buzzdeee Would it make sense to default $ruby_interpreter to something else when osfamily=OpenBSD ?

I don't know OpenBSD though, so maybe not.  There's a rubyversion fact available if that helps.

buzzdeee commented:
based on the feedback, the following changes:
 * just always use str2bool, instead of testing with is_bool, makes it much cleaner, and allows to keep stdlib version dependency to 4.0.0
 * for the ruby_interpreter in the OpenBSD case, use regsubst, to create a sane default based on the ::rubyversion fact
 * for the site_libdir, use the same approach, for OpenBSD, use regsubst, to point to the right directory

And some added/updated spec tests.

hope that's better now, let me know if there is still stuff improvable. ;)

alexjfisher commented:
Thanks for the updates.  OpenBSD support will be great to have.

Personally, I'm still uneasy about turning parameters that should be bools into ones that also accept strings just to overcome a limitation when using hiera with recursive lookups (a fairly rare use-case?).  It might make the module harder to maintain.
There are quite a few other boolean parameters.  Which of these should also be changed?

In newer versions of hiera, you might be able to use the alias function?
https://docs.puppetlabs.com/hiera/3.0/variables.html#the-alias-lookup-function

I checked through a bunch of puppetlabs supported modules to see what they did.  The majority call validate_bool for all their boolean parameters.  That's probably what I'd like to see this module doing too.

But, I'm really only expressing my own view.  I (rightly) have no commit or merge access and am not the person you'd need to convince.

buzzdeee commented:
Just to show how my setup is:
 * have a central variable defined, i.e.:
role::networkmanagement::puppetmaster::mcollective::middleware_ssl: true

then when configuring the middleware, I re-use that variable, as well as for the mcollective::users I have. So when I want to switch ssl between true and false, I only need to switch in that single place.

Anyways, if it really has to be, then I could also live without that convenicence ;) 

The Hiera version that comes with OpenBSD currently is 1.3.4. The maintainer once tried to update, but IIRC, there were issues with compatibility to the puppet version.

alexjfisher commented:
yeah, I think you need puppet 4 for hiera >=2.
I wonder if the yaml backend could be modified to support casting?
https://en.wikipedia.org/wiki/YAML#Casting_data_types

ffrank commented:
Puppet 4 will allow shenanigans such as `Variant[Boolean, Regex[/foo|bar|baz/]]`, so I guess future modules can be more flexible.

Not sure how good an idea that is from a UX perspective, but the technical hurdle will get lower.

igalic commented:
yeah, imo, until puppet 4 allows you to specify your own types somewhere, and we have a library filled with those, it's gonna be pretty damn unusable :(

ffrank commented:
Ah but thats Developer Experience :) I was thinking more along the lines of "it's easy to grok that this parameter should be a boolean". Parameters that behave differently when given different types of values can add cognitive load.

I feel that it's hard to discuss this without conducting some UX studies first though. What I'm saying is that I'm basically impartial at this point.

jyaworski commented:
Hello:

Can this be squashed and rebased?

buzzdeee commented:
squashed and merged as requested, and I'm just hoping that the rubocop will now also be happy (:

igalic commented:
dear past @igalic, i'd like to point out that puppet now *does* support Types declarations, and that you yourself have created a module with numerous types: https://github.com/voxpupuli/puppet-tea
